chained_filter_controller/src/chained_filter.cpp

```diff
@@ -0,0 +1,167 @@
+      }
+      RCLCPP_INFO_EXPRESSION(
+        get_node()->get_logger(), has_global_filter_setting,
+        "Using global filter setting for all interfaces.");
+    }
+
+    for (size_t i = 0; i < params_.input_interfaces.size(); ++i)
+    {
+      filters_[i] = std::make_unique<filters::FilterChain<double>>("double");
+      if (!filters_[i]->configure(
+            has_global_filter_setting ? "filter_chain"
+                                      : params_.input_interfaces[i] + ".filter_chain",
+            get_node()->get_node_logging_interface(), get_node()->get_node_parameters_interface()))
+      {
+        RCLCPP_ERROR(
+          get_node()->get_logger(),
+          "Failed to configure filter chain for interfaces  %s. Check the parameters for filters "
+          "setup.",
+          params_.input_interfaces[i].c_str());
+        return controller_interface::CallbackReturn::FAILURE;
+      }
+    }
+  }
+  catch (const std::exception & e)
+  {
+    fprintf(stderr, "Exception thrown during configure stage with message: %s \n", e.what());
+    return controller_interface::CallbackReturn::ERROR;
+  }
+  return controller_interface::CallbackReturn::SUCCESS;
+}
+
+controller_interface::CallbackReturn ChainedFilter::on_activate(const rclcpp_lifecycle::State &)
+{
+  std::fill(
+    output_state_values_.begin(), output_state_values_.end(),
+    std::numeric_limits<double>::quiet_NaN());
+  return controller_interface::CallbackReturn::SUCCESS;
+}
+
+controller_interface::return_type ChainedFilter::update_and_write_commands(
+  const rclcpp::Time &, const rclcpp::Duration &)
+{
+  for (size_t i = 0; i < params_.input_interfaces.size(); ++i)
+  {
+    const auto sensor_op = state_interfaces_[i].get_optional();
+    if (!sensor_op.has_value())
+    {
+      RCLCPP_ERROR(
+        get_node()->get_logger(), "Failed to read sensor value from state interface '%s'.",
+        state_interfaces_[0].get_name().c_str());
+      continue;
+    }
+    if (!std::isnan(sensor_op.value()))
+    {
+      filters_[i]->update(sensor_op.value(), output_state_values_[i]);
+    }
+  }
+
+  return controller_interface::return_type::OK;
+}
+
+std::vector<hardware_interface::StateInterface> ChainedFilter::on_export_state_interfaces()
+{
+  std::vector<hardware_interface::StateInterface> state_interfaces;
+  for (size_t i = 0; i < params_.output_interfaces.size(); ++i)
+  {
+    state_interfaces.emplace_back(
+      get_node()->get_name(), params_.output_interfaces.at(i), &output_state_values_.at(i));
+  }
+  return state_interfaces;
+}
+
+controller_interface::return_type ChainedFilter::update_reference_from_subscribers(
+  const rclcpp::Time &, const rclcpp::Duration &)
+{
+  return controller_interface::return_type::OK;
+}
+
+rclcpp::NodeOptions ChainedFilter::define_custom_node_options() const
+{
+  return rclcpp::NodeOptions()
+    .allow_undeclared_parameters(true)
+    .automatically_declare_parameters_from_overrides(false);
+}
```

Comment:
Won't it work with the default settings?, Just curious

Reply:
haven't tested it before, but it seems to work without